src/views/ChallengeLinkLanding/ChallengeLinkLanding.tsx

```diff
@@ -0,0 +1,229 @@
+import { errorAlerter } from "misc";
+import { browserHistory } from "ogsHistory";
+import { get_ebi } from "SignIn";
+import cached from "cached";
+import * as dynamic_help from "dynamic_help_config";
+import { Card } from "material";
+import { LoadingPage } from "Loading";
+import { Player } from "Player";
+import { ChallengeDetailsReviewPane } from "ChallengeDetailsReviewPane";
+
+type Challenge = socket_api.seekgraph_global.Challenge;
+
+// Users are intended to arrive here via a challenge-link URL - those point here.
+// This page will display a bouncing OGS logo until the challenge given in search param is loaded.
+// From there, the user can accept the game - going via login orn auto-registration if necessary.
+
+export function ChallengeLinkLanding(): JSX.Element {
+    const user = data.get("user");
+    const logged_in = !user.anonymous;
+
+    /* State */
+    const [linked_challenge, set_linked_challenge] = React.useState<Challenge>(null);
+    const [ask_to_login, set_ask_to_login] = React.useState<boolean>(false);
+
+    /* Actions */
+
+    const doAcceptance = (challenge: Challenge) => {
+        swal({
+            text: "Accepting game...",
+            type: "info",
+            showCancelButton: false,
+            showConfirmButton: false,
+            allowEscapeKey: false,
+        }).catch(swal.noop);
+
+        post("challenges/%%/accept", challenge.challenge_id, {})
+            .then(() => {
+                swal.close();
+                browserHistory.push(`/game/${challenge.game_id}`);
+            })
+            .catch((err) => {
+                swal.close();
+                errorAlerter(err);
+            });
+        data.set("pending_accepted_challenge", null);
+    };
+
+    const accept = () => {
+        if (logged_in) {
+            doAcceptance(linked_challenge);
+        } else {
+            set_ask_to_login(true);
+            // We need to save the challenge info in this way for when we come back after logging in.
+            // An alternative is passing this "state" via the URL hash or params, but believe me
+            // this does not work out well :)
+            data.set("pending_accepted_challenge", linked_challenge);
+            console.log("pended", linked_challenge);
+        }
+    };
+
+    const registerAndCommence = () => {
+        const new_username_root = pgettext(
+            "This needs to be a no-spaces honorific for a guest user",
+            "HonouredGuest",
+        );
+        // naively uniquify... 6 lsbs of epoch... good enough?
+        const new_username =
+            new_username_root.replace(/\s+/g, "") + Date.now().toString().slice(-6);
+
+        const initial_password = Date.now().toString(); // They will have to change this, so anything random & unique is OK
+
+        post("/api/v0/register", {
+            username: new_username,
+            password: initial_password,
+            email: "",
+            ebi: get_ebi(),
+        })
```

Comment:
I'm thinking we might want to move the username generation down to python so that we can quickly loop and look for a username that's not taken. I'm thinking basically if a username isn't set, or `generate_username` is set, or something like that, generate one in python, that way we can check for duplicates and regenerate as much as we need. Having 6 digits is already pretty long for a user to handle, but at the same time that's only enough space for a million users which we'll exhaust at some point and even with only a few thousand entries the likely hood of a collision happening to someone is pretty high I think, so best to deal with the regeneration as needed.

Reply:
Supposing we put it on the backend ... is it OK for it to just count up from 1?    Typically this is "not the done thing" for identifiers, but actually I like it - much like with games, it lets the users see how many there are, which is amusing.

Reply:
Sure that'd be fine by me